tools/schemacode/bidsschematools/validator.py

```diff
@@ -38,285 +154,226 @@ def _get_paths(
-            pattern = f"(?:{pattern})?"
-        dir_regex += pattern
-
-    # If we move to referring to suffixes by keys in the object table:
-    # suffixes = [schema.objects.suffixes[suffix].value for suffix in rule["suffixes"]]
-    suffixes = rule["suffixes"]
-    suffix_regex = f"(?P<suffix>{'|'.join(suffixes)})"
-
-    # If we move to referring to extensions by keys in the object table:
-    # extensions = [schema.objects.extensions[ext].value for ext in rule["extensions"]]
-    extensions = rule["extensions"]
-    ext_match = "|".join(_sanitize_extension(ext) for ext in extensions)
-    ext_regex = f"(?P<extension>{ext_match})"
-
-    return {
-        "regex": "".join(dir_regex + entity_regex + [suffix_regex, ext_regex]),
-        "mandatory": False,
-    }
-
-
-def load_filename_rules(
-    rule_group: bst.types.Namespace,
-    schema: bst.types.Namespace,
-    level: int,
-):
-    """Load schema rules into regular expressions
+    Raise errors for validation result.
 
     Parameters
     ----------
-    rule_group : Namespace
-        The set of rules to load from the schema
-    schema : Namespace
-        A nested dictionary, as returned by `bidsschematools.schema.load_schema()`.
-    level : int
-        The depth in rule_group to look for rules
-
-    Returns
-    -------
-    rules : list of dict
-        A list of dictionaries, with keys including 'regex' and 'mandatory'.
+    validation_result : dict
+        A dictionary as returned by `validate_all()` with keys including "schema_tracking",
+        "path_tracking", "path_listing", and, optionally "itemwise".
+        The "itemwise" value, if present, should be a list of dictionaries, with keys including
+        "path", "regex", and "match".
     """
-    regex_schema = []
-    for rule_template in rule_group.values(level=level):
-        # Simple rules, e.g. dataset_description.json, README
-        if "path" in rule_template:
-            regex_schema.append(_path_rule(rule_template))
-        elif "stem" in rule_template:
-            regex_schema.append(_stem_rule(rule_template))
-        else:
-            regex_schema.extend(
-                _entity_rule(rule, schema) for rule in split_inheritance_rules(rule_template)
+    total_file_count = len(validation_result["path_listing"])
+    validated_files_count = total_file_count - len(validation_result["path_tracking"])
+    if validated_files_count == 0:
+        lgr.error("No valid BIDS files were found.")
+    for entry in validation_result["schema_tracking"]:
+        if entry["mandatory"]:
+            lgr.error(
+                "The `%s` regex pattern file required by BIDS was not found.",
+                entry["regex"],
             )
-
-    return regex_schema
+    for i in validation_result["path_tracking"]:
+        lgr.warning("The `%s` file was not matched by any regex schema entry.", i)
 
 
-@lru_cache()
-def load_all(
-    schema_dir,
+def select_schema_path(
+    bids_version="",
+    bids_root="",
```

Comment:
It's generally best to use a sentinel value like `None` instead of a degenerate value like `""` for optional parameters. That more clearly indicates that the default behavior is in-kind different from the behavior when the thing is specified.

In contrast, if expecting a sequence, you can use an empty sequence like `()` as the default value because you will iterate over any sequence passed.